### DISCUSSION TITLE
Gerenciar pacotes instalados globalmente

### DISCUSSION THREAD

**Discussion by tapmorales** [frontendbr/forum#805] [Perguntas e Respostas]
Como voces fazem para gerenciar os pacotes instalados pelo npm para não ter que instala-los globalmente?

Sei que tem pacotes do npm para gerenciar as versoes destes pacotes, mas gostaria de saber como vocês lidam com isso, ou instalam globalmente mesmo quando necessário.

**Comment by fdaciuk** [comment]
Eu tento instalar o mínimo possível de pacotes globais, só alguns realmente específicos.
No mais, eu tento manter as dependências por projeto, instalando como dependência de desenvolvimento, o que poderia ser global.

Como eu uso linux, eu configuro o `$PATH` do meu sistema pra usar qualquer ferramenta de CLI que esteja no projeto direto na linha de comando, sem precisar criar um _NPM Script_ pra isso.

A configuração seria essa:

No arquivo de configuração do seu shell, adicione no final dele (se você usa o bash, provavelmente vai ser o `.bashrc`, como eu uso ZSH, o arquivo é o `.zshrc`, que fica na raiz do diretório `home`: `~/.zshrc`):

```zsh
export PATH="node_modules/.bin:$PATH"
```

Feche e abra novamente seu terminal, ou execute `source ~/.zshrc` para pegar a última alteração.

Fazendo isso, o `node_modules/.bin` local será adicionado ao `$PATH` do sistema. Quando você estiver dentro de um projeto que tenha uma CLI instalada via NPM, na raiz do seu projeto você vai ter o diretório `node_modules/.bin/<nome-do-executavel>`.

Como o `node_modules/.bin` foi adicionado ao `$PATH`, quando você rodar um comando no terminal, esse caminho vai ser usado pelo SO para verificar se há algum executável com esse nome.

Por exemplo: instalando o `webpack` como _devDependency_, na `node_modules/.bin` vai ter o executável do `webpack`. Ao rodar o comando `webpack` no seu terminal, ele vai funcionar normalmente :)

Acho que me repeti bastante, mas é isso :stuck_out_tongue:

**Comment by fdaciuk** [comment]
Outro ponto importante de citar é que, ao executar qualquer NPM script usando alguma CLI, o NPM já faz essa busca em `node_modules/.bin`, então se não quiser fazer o que falei acima, você pode adicionar um NPM Script no seu projeto para rodar a CLI.

No `package.json` do seu projeto, seria somente adicionar na entrada `scripts`:

```
{
  "scripts": {
    "webpack": "webpack"
  }
}
```
E, ao executar `npm run webpack` (ou `yarn webpack`, se usar o Yarn), o comando `webpack` vai ser executado à partir da sua `node_modules/.bin` :)

**Comment by doug2k1** [comment]
Outra forma de executar a partir do `node_modules/.bin` sem precisar de npm script é usar o [npx](http://blog.npmjs.org/post/162869356040/introducing-npx-an-npm-package-runner) (vem junto com o npm 5.2.0+).

É só usar `npx webpack`. Ele primeiro vai procurar globalmente, se não existir, ele procura no node_modules local. Se ainda não achar, ele instala na hora antes de executar.

**Comment by tapmorales** [comment]
Obrigado gente,

Eu uso o windows, pra resolver a questão de versões do node em projetos diferentes tem o nvm.

Mas como eu estou começando a estudar o Ionic (fazendo testes mesmos) e já tive problemas com phonegap, queria saber como lidar com isso. O Phonegap pede pra instalar globalmente. A maioria dos pacotes que instalei na vida eu instalo local e uso o npm run para rodar o executavel local. Mas isso nao funcionou com o phonegap. O problema é que se eu tenho mais de um app na minha maquina e pra um deles eu preciso atualizar o phonegap global, corro o risco de quebrar um app que só funciona numa versão mais antiga.

Não conhecia o npx, valeu @doug2k1 .

Valeu pela aula @fdaciuk hehehe. Só uma coisa, todos na equipe tem que adotar esse padrão do PATH. Não vejo problemas, é um procedimento que todos devem seguir.
Contudo, digamos que voce tenha o projetoA e o projetoB. O projetoA usa a dependencia na versao1. O B usa na versao2. No PATH vc tem os dois caminhos configurados: /home/projetoA/node_modules/ e /home/projetoB/node_modules/. Como fazer para que quando digitar no terminal <dependencia> dentro de projetoA, use a versao 1, e quando estiver no ProjetoB, use a versao 2?

**Comment by fdaciuk** [comment]
> Contudo, digamos que voce tenha o projetoA e o projetoB. O projetoA usa a dependencia na versao1. O B usa na versao2. No PATH vc tem os dois caminhos configurados: /home/projetoA/node_modules/ e /home/projetoB/node_modules/. Como fazer para que quando digitar no terminal dentro de projetoA, use a versao 1, e quando estiver no ProjetoB, use a versao 2?

Então, não faz! Por isso você coloca só `node_modules/.bin` no `$PATH`. Ele assume que, onde você vai executar o comando, a `node_modules` tem que estar no mesmo nível, logo funciona pra qualquer projeto, desde que você execute o comando na raiz do projeto :)

**Comment by tapmorales** [comment]
Boa. Entendi. Valeu.

Alias, excelente ideia. Curti
